Tasks/Gradle/CodeAnalysis/SonarQube/metrics.ts

```diff
@@ -98,51 +172,84 @@ export class SonarQubeMetrics {
     }
 
     /**
+     * For all the units used by the server, this method returns the key and the friendly name, as well as the type and the id
+     * @returns {Promise<Object>} A list of all units used by the SQ server, represented by objects with fields: id, key, type, name. Rejects if server response was invalid.
+     */
+    public getMeasurementDetails():Q.Promise<SonarQubeMeasurementUnit[]> {
+        if (SonarQubeMetrics.isCached(this.measurementUnits)) {
+            tl.debug(`[SQ] Measurement units cache hit (cached array length ${this.measurementUnits.length})`);
+            return Q.when(this.measurementUnits);
+        }
+
+        return this.server.invokeApiCall('/api/metrics/search?ps=500&f=name')
+            .then((response:any) => {
+                return Q.Promise<SonarQubeMeasurementUnit[]>((resolve, reject) => {
+                    if ((response == undefined || response == null) ||
+                        (response.metrics == undefined || response.metrics == null)) {
+                        reject(new Error(tl.loc('sqAnalysis_NoUnitsFound')));
+                    } else {
+                        this.measurementUnits = response.metrics;
+                        resolve(Q.when(this.measurementUnits));
+                    }
+
+                });
+            });
+    }
+
+    /**
      * Repeatedly query the server to determine if the task has finished.
      * @param timeout    (optional) Time, in seconds, to wait for the analysis job to finish before returning the promise as false.
      * @param delay      (optional) Time, in seconds, to wait between polling attempts. Warning: Does not also wait for previous request to finish.
      * @returns A promise resolving true if the task finished or rejecting if there was an error or the timeout was exceeded.
      */
     private waitForTaskCompletion(timeout?:number, delay?:number):Q.Promise<boolean> {
-        tl.debug(`[SQ] Waiting for SonarQube analysis to complete `);
+        tl.debug(`[SQ] Waiting for SonarQube analysis to complete.`);
 
         // If we have previously waited, then we can immediately return true.
         if (this.analysisComplete) {
-            tl.debug(`[SQ] Analysis complete check already performed `);
+            tl.debug(`[SQ] Analysis already cached as complete.`);
             return Q.when(true);
         }
 
+        // Call isTaskComplete immediately
```

Comment:
This section has been changed to avoid a problem that previously existed where the first request to the server would not be sent until after one delay period. Now a request is immediately fired off before setting up the repeating tasks.